File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartFullScreenDialogFragment.java
package org.wordpress.android.ui.quickstart;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import org.wordpress.android.R;
import org.wordpress.android.WordPress;
import org.wordpress.android.analytics.AnalyticsTracker;
import org.wordpress.android.analytics.AnalyticsTracker.Stat;
import org.wordpress.android.fluxc.store.QuickStartStore;
import org.wordpress.android.fluxc.store.QuickStartStore.QuickStartTask;
import org.wordpress.android.fluxc.store.QuickStartStore.QuickStartTaskType;
import org.wordpress.android.ui.ActionableEmptyView;
import org.wordpress.android.ui.FullScreenDialogFragment.FullScreenDialogContent;
import org.wordpress.android.ui.FullScreenDialogFragment.FullScreenDialogController;
import org.wordpress.android.ui.prefs.AppPrefs;
import org.wordpress.android.ui.quickstart.QuickStartAdapter.OnQuickStartAdapterActionListener;
import org.wordpress.android.util.AniUtils;
import org.wordpress.android.util.AniUtils.Duration;
import org.wordpress.android.util.QuickStartUtils;

import java.util.ArrayList;
import java.util.List;

import javax.inject.Inject;

import static org.wordpress.android.fluxc.store.QuickStartStore.QuickStartTaskType.CUSTOMIZE;
import static org.wordpress.android.fluxc.store.QuickStartStore.QuickStartTaskType.GROW;

public class QuickStartFullScreenDialogFragment extends Fragment implements FullScreenDialogContent,
        OnQuickStartAdapterActionListener {
    private FullScreenDialogController mDialogController;
    private QuickStartAdapter mQuickStartAdapter;
    private ActionableEmptyView mQuickStartCompleteView;

    public static final String KEY_COMPLETED_TASKS_LIST_EXPANDED = "completed_tasks_list_expanded";
    public static final String EXTRA_TYPE = "EXTRA_TYPE";
    public static final String RESULT_TASK = "RESULT_TASK";

    private QuickStartTaskType mTasksType = CUSTOMIZE;

    @Inject protected QuickStartStore mQuickStartStore;

    public static Bundle newBundle(QuickStartTaskType type) {
        Bundle bundle = new Bundle();
        bundle.putSerializable(EXTRA_TYPE, type);
        return bundle;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ((WordPress) requireActivity().getApplication()).component().inject(this);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {
        ViewGroup rootView = (ViewGroup) inflater.inflate(R.layout.quick_start_dialog_fragment, container, false);

        if (getArguments() != null) {
            mTasksType = (QuickStartTaskType) getArguments().getSerializable(EXTRA_TYPE);
        }

        RecyclerView list = rootView.findViewById(R.id.list);
        List<QuickStartTask> tasksUncompleted = new ArrayList<>();
        List<QuickStartTask> tasksCompleted = new ArrayList<>();
        int site = AppPrefs.getSelectedSite();

        mQuickStartCompleteView = rootView.findViewById(R.id.quick_start_complete_view);

        switch (mTasksType) {
            case CUSTOMIZE:
                tasksUncompleted.addAll(mQuickStartStore.getUncompletedTasksByType(site, CUSTOMIZE));
                tasksCompleted.addAll(mQuickStartStore.getCompletedTasksByType(site, CUSTOMIZE));
                setCompleteViewImage(R.drawable.img_illustration_site_brush_191dp);
                AnalyticsTracker.track(Stat.QUICK_START_TYPE_CUSTOMIZE_VIEWED);
                break;
            case GROW:
                tasksUncompleted.addAll(mQuickStartStore.getUncompletedTasksByType(site, GROW));
                tasksCompleted.addAll(mQuickStartStore.getCompletedTasksByType(site, GROW));
                setCompleteViewImage(R.drawable.img_illustration_site_about_182dp);
                AnalyticsTracker.track(Stat.QUICK_START_TYPE_GROW_VIEWED);
                break;
            case UNKNOWN:
                tasksUncompleted.addAll(mQuickStartStore.getUncompletedTasksByType(site, CUSTOMIZE));
                tasksCompleted.addAll(mQuickStartStore.getCompletedTasksByType(site, CUSTOMIZE));
                setCompleteViewImage(R.drawable.img_illustration_site_brush_191dp);
                break;
        }

        boolean isCompletedTasksListExpanded = savedInstanceState != null
                                               && savedInstanceState.getBoolean(KEY_COMPLETED_TASKS_LIST_EXPANDED);

        mQuickStartAdapter = new QuickStartAdapter(
                requireContext(),
                tasksUncompleted,
                tasksCompleted,
                isCompletedTasksListExpanded);

        if (tasksUncompleted.isEmpty()) {
            mQuickStartCompleteView.setVisibility(!isCompletedTasksListExpanded ? View.VISIBLE : View.GONE);
        }

        mQuickStartAdapter.setOnTaskTappedListener(QuickStartFullScreenDialogFragment.this);
        list.setLayoutManager(new LinearLayoutManager(requireContext()));
        list.setAdapter(mQuickStartAdapter);
        // Disable default change animations to avoid blinking effect when adapter data is changed.
        ((DefaultItemAnimator) list.getItemAnimator()).setSupportsChangeAnimations(false);

        return rootView;
    }

    @Override
    public void setController(final FullScreenDialogController controller) {
        mDialogController = controller;
    }

    @Override
    public boolean onConfirmClicked(FullScreenDialogController controller) {
        return true;
    }

    @Override
    public boolean onDismissClicked(FullScreenDialogController controller) {
        switch (mTasksType) {
            case CUSTOMIZE:
                AnalyticsTracker.track(Stat.QUICK_START_TYPE_CUSTOMIZE_DISMISSED);
                break;
            case GROW:
                AnalyticsTracker.track(Stat.QUICK_START_TYPE_GROW_DISMISSED);
                break;
            case UNKNOWN:
                // Do not track unknown.
                break;
        }

        controller.dismiss();
        return true;
    }

    @Override
    public void onTaskTapped(QuickStartTask task) {
        AnalyticsTracker.track(QuickStartUtils.getQuickStartListTappedTracker(task));
        Bundle result = new Bundle();
        result.putSerializable(RESULT_TASK, task);
        mDialogController.confirm(result);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        if (mQuickStartAdapter != null) {
            outState.putBoolean(KEY_COMPLETED_TASKS_LIST_EXPANDED, mQuickStartAdapter.isCompletedTasksListExpanded());
        }
    }

    @Override
    public void onSkipTaskTapped(QuickStartTask task) {
        AnalyticsTracker.track(QuickStartUtils.getQuickStartListSkippedTracker(task));
        mQuickStartStore.setDoneTask(AppPrefs.getSelectedSite(), task, true);

        if (mQuickStartAdapter != null) {
            int site = AppPrefs.getSelectedSite();

            List<QuickStartTask> uncompletedTasks = mQuickStartStore.getUncompletedTasksByType(site, mTasksType);

            mQuickStartAdapter.updateContent(
                    uncompletedTasks,
                    mQuickStartStore.getCompletedTasksByType(site, mTasksType));

            if (uncompletedTasks.isEmpty() && !mQuickStartAdapter.isCompletedTasksListExpanded()) {
                toggleCompletedView(true);
            }
        }
    }

    @Override
    public void onCompletedTasksListToggled(boolean isExpanded) {
        switch (mTasksType) {
            case CUSTOMIZE:
                AnalyticsTracker.track(isExpanded ? Stat.QUICK_START_LIST_CUSTOMIZE_EXPANDED
                        : Stat.QUICK_START_LIST_CUSTOMIZE_COLLAPSED);
                break;
            case GROW:
                AnalyticsTracker.track(isExpanded ? Stat.QUICK_START_LIST_GROW_EXPANDED
                        : Stat.QUICK_START_LIST_GROW_COLLAPSED);
                break;
            case UNKNOWN:
                // Do not track unknown.
                break;
        }

        if (mQuickStartStore.getUncompletedTasksByType(AppPrefs.getSelectedSite(), mTasksType).isEmpty()) {
            toggleCompletedView(!isExpanded);
        }
    }

    private void setCompleteViewImage(int imageResourceId) {
        mQuickStartCompleteView.image.setImageResource(imageResourceId);
        mQuickStartCompleteView.image.setVisibility(View.VISIBLE);
    }

    private void toggleCompletedView(boolean isVisible) {
        if (isVisible) {
            AniUtils.fadeIn(mQuickStartCompleteView, Duration.SHORT);
        } else {
            AniUtils.fadeOut(mQuickStartCompleteView, Duration.SHORT);
        }
    }
}

File: WordPress/src/main/java/org/wordpress/android/ui/quickstart/QuickStartFullScreenDialogFragment.kt
package org.wordpress.android.ui.quickstart

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.recyclerview.widget.DefaultItemAnimator
import androidx.recyclerview.widget.LinearLayoutManager
import com.google.android.material.snackbar.Snackbar
import org.wordpress.android.R
import org.wordpress.android.WordPress
import org.wordpress.android.databinding.QuickStartDialogFragmentBinding
import org.wordpress.android.fluxc.store.QuickStartStore
import org.wordpress.android.fluxc.store.QuickStartStore.QuickStartNewSiteTask.CREATE_SITE
import org.wordpress.android.fluxc.store.QuickStartStore.QuickStartTask
import org.wordpress.android.fluxc.store.QuickStartStore.QuickStartTaskType
import org.wordpress.android.fluxc.store.QuickStartStore.QuickStartTaskType.UNKNOWN
import org.wordpress.android.ui.FullScreenDialogFragment.FullScreenDialogContent
import org.wordpress.android.ui.FullScreenDialogFragment.FullScreenDialogController
import org.wordpress.android.ui.mysite.SelectedSiteRepository
import org.wordpress.android.ui.mysite.cards.quickstart.QuickStartCardBuilder
import org.wordpress.android.ui.quickstart.QuickStartFullScreenDialogFragment.QuickStartListCard.QuickStartHeaderCard
import org.wordpress.android.ui.quickstart.QuickStartFullScreenDialogFragment.QuickStartListCard.QuickStartTaskCard
import org.wordpress.android.ui.utils.UiHelpers
import org.wordpress.android.ui.utils.UiString
import org.wordpress.android.ui.utils.UiString.UiStringRes
import org.wordpress.android.util.DisplayUtilsWrapper
import org.wordpress.android.util.QuickStartUtils.getQuickStartListSkippedTracker
import org.wordpress.android.util.QuickStartUtils.getQuickStartListTappedTracker
import org.wordpress.android.widgets.WPSnackbar.Companion.make
import java.io.Serializable
import javax.inject.Inject

class QuickStartFullScreenDialogFragment : Fragment(R.layout.quick_start_dialog_fragment),
        FullScreenDialogContent {
    private var dialogController: FullScreenDialogController? = null
    private var tasksType: QuickStartTaskType = QuickStartTaskType.CUSTOMIZE
    private lateinit var quickStartAdapter: QuickStartAdapter

    @Inject lateinit var quickStartTracker: QuickStartTracker
    @Inject lateinit var quickStartStore: QuickStartStore
    @Inject lateinit var selectedSiteRepository: SelectedSiteRepository
    @Inject lateinit var uiHelpers: UiHelpers
    @Inject lateinit var quickStartCardBuilder: QuickStartCardBuilder
    @Inject lateinit var displayUtilsWrapper: DisplayUtilsWrapper

    private var _binding: QuickStartDialogFragmentBinding? = null
    private val binding get() = _binding!!

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        (requireActivity().application as WordPress).component().inject(this)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        _binding = QuickStartDialogFragmentBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        tasksType = arguments?.getSerializable(EXTRA_TYPE) as QuickStartTaskType? ?: QuickStartTaskType.UNKNOWN
        quickStartTracker.trackQuickStartListViewed(tasksType)
        binding.setupQuickStartList()
    }

    private fun QuickStartDialogFragmentBinding.setupQuickStartList() {
        quickStartAdapter = QuickStartAdapter(uiHelpers)
        list.layoutManager = LinearLayoutManager(requireContext())
        list.adapter = quickStartAdapter
        // Disable default change animations to avoid blinking effect when adapter data is changed.
        (list.itemAnimator as DefaultItemAnimator?)?.supportsChangeAnimations = false
        updateQuickStartList()
    }

    private fun updateQuickStartList() {
        val quickStartList = mutableListOf<QuickStartListCard>().apply {
            add(buildHeaderCard())
            addAll(buildTaskCards())
        }.toList()
        quickStartAdapter.submitList(quickStartList)
    }

    override fun setController(controller: FullScreenDialogController) {
        dialogController = controller
    }

    override fun onConfirmClicked(controller: FullScreenDialogController): Boolean {
        return true
    }

    override fun onDismissClicked(controller: FullScreenDialogController): Boolean {
        quickStartTracker.trackQuickStartListDismissed(tasksType)
        controller.dismiss()
        return true
    }

    private fun onTaskTapped(task: QuickStartTask) {
        quickStartTracker.track(getQuickStartListTappedTracker(task))
        if (!showSnackbarIfNeeded(task)) {
            val result = Bundle()
            result.putSerializable(RESULT_TASK, task as Serializable?)
            dialogController?.confirm(result)
        }
    }

    private fun onSkipTaskTapped(task: QuickStartTask) {
        quickStartTracker.track(getQuickStartListSkippedTracker(task))
        val selectedSiteLocalId = selectedSiteRepository.getSelectedSiteLocalId().toLong()
        quickStartStore.setDoneTask(selectedSiteLocalId, task, true)
        updateQuickStartList()
    }

    private fun showSnackbarIfNeeded(task: QuickStartTask?): Boolean {
        return if (task === CREATE_SITE) {
            make(
                    requireView(),
                    R.string.quick_start_list_create_site_message,
                    Snackbar.LENGTH_LONG
            ).show()
            true
        } else {
            false
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _binding = null
    }

    private fun buildHeaderCard() = QuickStartHeaderCard(
            title = UiStringRes(quickStartCardBuilder.getTitle(tasksType)),
            shouldShowHeaderImage = !displayUtilsWrapper.isPhoneLandscape()
    )

    private fun buildTaskCards(): List<QuickStartTaskCard> {
        val tasks = QuickStartTask.getTasksByTaskType(tasksType).filterNot { it.taskType == UNKNOWN }
        val selectedSiteLocalId = selectedSiteRepository.getSelectedSiteLocalId().toLong()
        val tasksCompleted = quickStartStore.getCompletedTasksByType(selectedSiteLocalId, tasksType)
        return tasks.mapToQuickStartTaskCard(tasksCompleted)
    }

    private fun List<QuickStartTask>.mapToQuickStartTaskCard(tasksCompleted: List<QuickStartTask>) = this.map {
        QuickStartTaskCard(
                task = it,
                isCompleted = tasksCompleted.contains(it),
                onTaskTapped = this@QuickStartFullScreenDialogFragment::onTaskTapped,
                onSkipTaskTapped = this@QuickStartFullScreenDialogFragment::onSkipTaskTapped
        )
    }

    sealed class QuickStartListCard {
        data class QuickStartHeaderCard(
            val title: UiString,
            val shouldShowHeaderImage: Boolean
        ) : QuickStartListCard()

        data class QuickStartTaskCard(
            val task: QuickStartTask,
            val isCompleted: Boolean,
            val onTaskTapped: (task: QuickStartTask) -> Unit,
            val onSkipTaskTapped: (task: QuickStartTask) -> Unit
        ) : QuickStartListCard()
    }

    companion object {
        const val EXTRA_TYPE = "EXTRA_TYPE"
        const val RESULT_TASK = "RESULT_TASK"
        fun newBundle(type: QuickStartTaskType?): Bundle {
            val bundle = Bundle()
            bundle.putSerializable(EXTRA_TYPE, type)
            return bundle
        }
    }
}
